juju/juju
# [JUJU-2340] Fixes upgrade bugs for sidecar charms in 2.9

In the process of developing 2.9 we have made the mistake of placing pod spec changes into the sidecar charms ensure method meant for newer versions of Juju.

With this what ends up happening is that when a user upgrades a controller the ensure method runs again for the sidecar application and try's to start overwriting the pod spec with new addition's not currently intended for the version of juju the model is running.

With this change we have placed these changes by gate's to check agent versions when ensuring. We have also removed the upgrade path for sidecard charms and moved the logic back to ensuring the state we expect for different agent versions.

Fixes lp1997253

## Checklist


## QA steps

To verify this fix we need to do an upgrade with Juju. The easiest way to do this is to follow the conditions described in the bug.

1. Deploy a juju 2.9.34 controller into a microk8s or minikube cluster. Make sure to set `config --caas-image-repo=<docker_username>` as part of the bootstrap.
2. Add a new model
3. Deploy snappass-test
4. Build the oci image for this PR and the juju client and push into minikube.
```sh
DOCKER_USERNAME=<docker_username> make juju minikube-operator-update
```
5. Perform the controller upgrade with `juju upgrade-controller --agent-stream=develop`
6. We now need to confirm that after controller upgrade the snappass-test pods have not gone into a crash state. We expect to see that the charm container may have restarted during the upgrade. This is happening due it loosing connection with the controller and the liveliness probes restart the container.
7. Now we need to upgrade the model with `juju upgrade-model`. We want to confirm that after the upgrade the statefulset  for snappass-test has been updated. This can be done by confirming the charm container is now being run with pebble.

If all pods come back healthy after both upgrades then this is a successful test.

## Documentation changes

N/A

## Bug reference

lp1997253
